Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -0,0 +1,215 @@
+            // Storing this message could cause data corruption by fooling the cache into thinking the prior message is the last in the cache.
+            // JSON values (even empty ones) are always encoded with length > 0. If this condition is hit, the message is clearly corrupt.
+            throw CacheAdvanceWriteError.messageCorrupted
+        }
+
+        let bytesNeededToStoreMessage = messageLength + lengthOfMessageSuffix
+        guard bytesNeededToStoreMessage <= maximumBytes && bytesNeededToStoreMessage < Bytes(MessageSpan.max) else {
+            // The message is too long to be written to a cache of this size.
+            throw CacheAdvanceWriteError.messageDataTooLarge
+        }
+
+        let cacheHasSpaceForNewMessageBeforeEndOfFile = writer.offsetInFile + bytesNeededToStoreMessage <= maximumBytes
+        if shouldOverwriteOldMessages {
+            if !cacheHasSpaceForNewMessageBeforeEndOfFile {
+                // This message can't be written without exceeding our maximum file length.
+                // We'll need to start writing the file from the beginning of the file.
+
+                // Trim the file to the current position to remove soon-to-be-abandoned data from the file.
+                try writer.truncate(atOffset: writer.offsetInFile)
+
+                // Set the offset back to the beginning of the file.
+                try writer.seek(toOffset: 0)
+
+                // We know the oldest message is at the beginning of the file, since we just tossed out the rest of the file.
+                try reader.seek(toOffset: 0)
+
+                // We know we're about to overwrite the oldest message, so advance the reader to the second oldest message.
+                try reader.seekToNextMessage(shouldSeekToOldestMessageIfFound: false, cacheOverwritesOldMessages: true)
+            }
+
+            // Prepare the reader before writing the message.
+            try prepareReaderForWriting(dataOfLength: bytesNeededToStoreMessage)
+
+            // Create the marker for the offset representing the beggining of the message that will be the oldest once our write is done.
+            let offsetInFileOfOldestMessage = Data(Bytes(reader.offsetInFile))
+
+            // Write the message and oldest message offset.
+            try write(messageData: messageData, offsetInFileOfOldestMessage: offsetInFileOfOldestMessage)
+
+        } else if cacheHasSpaceForNewMessageBeforeEndOfFile {
+            // Write the message.
+            try write(messageData: messageData, offsetInFileOfOldestMessage: nil)
+
+        } else {
+            // We're out of room.
+            throw CacheAdvanceWriteError.messageDataTooLarge
+        }
+    }
+
+    /// Fetches all messages from the cache.
+    public func messages() throws -> [T] {
+        try setUpFileHandlesIfNecessary()
+
+        var messages = [T]()
+        while let encodedMessage = try reader.nextEncodedMessage(cacheOverwritesOldMessages: shouldOverwriteOldMessages) {
```

Comment:
> It seems like the reader is stateful.

It is! At the end of each `public` method call, the `reader` is at the beginning of the oldest known message.

> how is the reader being reset?

Because we read until the reader is back on the oldest known message 🙂

> invoking this twice in a row could be a good test

can't hurt! I'll add it.